### ISSUE TITLE
[subset] Expand the input unicode set to include decompositions and compositions.

### ISSUE THREAD

**Issue by garretrieger** [harfbuzz/harfbuzz#2283]
Since the shaper may automatically apply decompositions or compositions of characters even if there are no GSUB rules in a font for them. For a subset to be fully correct we should automatically expand the input unicode set to include these possible substitutions.

For example say we want to cut a subset for the text 0x41 (A), 0x30A (combining ring). For that text the shaper may choose to use 0xC5 (A with ring) instead of 0x41 + 0x30A. Today the subsetter will only retain glyphs for A and combining ring, which means that it'll be missing the glyph for A with ring.

I wrote a proof of concept implementation for a closure algorithm in java: https://gist.github.com/garretrieger/9b470871ae5c2159d2c083aa072d25b1. I need to port this to C++ to use in harfbuzz.

**Comment by khaledhosny** [comment]
> For that text the shaper may choose to use 0xC5 (A with ring) instead of 0x41 + 0x30A. Today the subsetter will only retain glyphs for A and combining ring, which means that it'll be missing the glyph for A with ring.

HarfBuzz will not apply composition if the composed glyph is missing from the font. So if the font has glyphs for only 0x41 + 0x30A, 0xC5 will be decomposed by 0x41 + 0x30A will never be composed.

**Comment by garretrieger** [comment]
Right, which can mean that the rendering for the subset will end up being different than for the whole font where the single combined glyph is used instead of the two separate glyphs.

A real example of this that we encountered was in Bengali for the font Hind Siliguri, the subset text was 0x09af + 0x09cc. With the whole font the shaper decomposes this to: 0x9c7 + 0x9af + 0x9d7 and everything renders correctly. When the font is subset to {0x9af, 0x9cc} currently the glyphs for 0x9c7 and 0x9d7 get dropped and then the subsetted font incorrectly renders.

**Comment by khaledhosny** [comment]
But isn’t this is a slightly different issue, as it is not canonical decomposition per se. I believe this is script shaping logic and does not always correspond to the normalization rules, and should be handled regardless of how normalization is handled.

**Comment by behdad** [comment]
> But isn’t this is a slightly different issue, as it is not canonical decomposition per se. I believe this is script shaping logic and does not always correspond to the normalization rules, and should be handled regardless of how normalization is handled.

Right. That knowledge needs to be reflected somewhere and used in both subsetters.

**Comment by behdad** [comment]
Decomposing is easy. For composing, we need a new callback on `hb_unicode_funcs_t` to close a set of unicode points over compositions.

**Comment by behdad** [comment]
I have a design in mind, which I need to write down, to use the shape-plan to drive subsetter glyph closure. That is what is needed proper.

**Comment by behdad** [comment]
> I have a design in mind, which I need to write down, to use the shape-plan to drive subsetter glyph closure. That is what is needed proper.

We experimented with shape-plan-driven closure in https://github.com/harfbuzz/harfbuzz/pull/3742. Unfortunately the results where slow and unsatisfactory :(. Without that would be hard to fix this issue.

**Comment by behdad** [comment]
I feel you...

Maybe we should close over compositions, but not decompositions?  We would need new API in the unicode-funcs though, which sucks...

@jfkthame thoughts?

**Comment by jfkthame** [comment]
That wouldn't fix things like the Bengali example Garret mentioned in https://github.com/harfbuzz/harfbuzz/issues/2283#issuecomment-605321453, though. I think if we're going to do this, we should do it fully. But I also note your "slow and unsatisfactory" comment at https://github.com/harfbuzz/harfbuzz/issues/2283#issuecomment-1233104104. :(

I don't see any details in https://github.com/harfbuzz/harfbuzz/pull/3742 about what the problems were with the approach attempted (and abandoned) there... do you remember why it turned into a dead end?

**Comment by behdad** [comment]
> That wouldn't fix things like the Bengali example Garret mentioned in [#2283 (comment)](https://github.com/harfbuzz/harfbuzz/issues/2283#issuecomment-605321453), though. I think if we're going to do this, we should do it fully. But I also note your "slow and unsatisfactory" comment at [#2283 (comment)](https://github.com/harfbuzz/harfbuzz/issues/2283#issuecomment-1233104104). :(

Decomposing is the easy part. The problem is when eg. an `Aacute` is decomposed to `A` and `acutecomb`, then we need to include all compositions of `acutecomb` with other chars as well, or the font will be "broken".
